env/env_encrypt2_impl.h

```diff
@@ -0,0 +1,81 @@
+//  copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#ifdef ROCKSDB_OPENSSL_AES_CTR
+#ifndef ROCKSDB_LITE
+
+#include "openssl/aes.h"
+#include "openssl/evp.h"
+#include "rocksdb/env_encrypt2.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// following define block from page 70:
+//  https://www.intel.com/content/dam/doc/white-paper/advanced-encryption-standard-new-instructions-set-paper.pdf
+#if !defined(ALIGN16)
+#if defined(__GNUC__)
+#define ALIGN16 __attribute__((aligned(16)))
+#else
+#define ALIGN16 __declspec(align(16))
+#endif
+#endif
+
+constexpr uint8_t kEncryptCodeVersion0{'0'};
+
+typedef char EncryptMarker[8];
+static EncryptMarker kEncryptMarker = "Encrypt";
+
+// long term:  code_version could be used in a switch statement or factory
+// prefix version 0 is 12 byte sha1 description hash, 128 bit (16 byte)
+// nounce (assumed to be packed/byte aligned)
+typedef struct {
+  uint8_t key_description_[EVP_MAX_MD_SIZE];  // max md is 64
+  uint8_t nonce_[AES_BLOCK_SIZE];             // block size is 16
+} PrefixVersion0;
```

Comment:
It seems like the header should have:
- The Marker that says it is encrypted
- The version of the header
- Maybe the type of CIPHER to use?
- The type of the description (sha1 vs md5 vs ???)
- The "description" (probably should called hash_description or something as it is not a text description)
- The nonce

Reply:
This is a design scope issue.  Nothing you suggest is wrong.  But I do not need it.  I put that '0' as code version at the end of the 'Encrypt' header to let anyone redefine what the next bytes contain.  'Tis why I call this branch "changeable encryption".  Someone could decide, 9 months into actual usage of this code, that any one or more of your variables are important and make code to change.  My explicit goal is to not anticipate future needs, but to create wiggle room to accommodate all of them.  AND, create a smooth transition path that does NOT require exporting and re-importing terabytes of data for an encryption design change.